File: References/Sunray/General_Module/sunray_tutorial/ugv_python/ugv_circle_vel.py
#!/usr/bin/env python3
import rospy
from sunray_msgs.msg import UGVControlCMD, UGVState
import math
import signal

class UGVCircleVel:
    def __init__(self):
        # 初始化参数
        self.ugv_id = rospy.get_param('~ugv_id', 1)
        self.linear_speed = rospy.get_param('~linear_speed', 0.5)            # 线速度
        self.circle_radius = rospy.get_param('~circle_radius', 1.0)          # 圆半径
        self.circle_center_x = rospy.get_param('~circle_center_x', 0.0)      # 圆中心x坐标
        self.circle_center_y = rospy.get_param('~circle_center_y', 0.0)      # 圆中心y坐标
        self.radial_gain = rospy.get_param('~radial_gain', 0.5)              # 径向增益
        self.initial_move_completed = False                         # 初始移动完成标志  
        self.returning_to_origin = False                            # 返回原点标志
        self.shutdown_requested = False                             # 关闭请求标志
        self.current_state = None                                   # 当前状态  
        self.INITIAL_MOVE_DISTANCE = 0.5                            # 初始移动距离

        # 发布控制指令和订阅状态
        self.cmd_pub = rospy.Publisher(f"/ugv{self.ugv_id}/sunray_ugv/ugv_control_cmd", UGVControlCMD, queue_size=10)
        self.state_sub = rospy.Subscriber(f"/ugv{self.ugv_id}/sunray_ugv/ugv_state", UGVState, self.state_callback)

        # 注册信号处理函数
        signal.signal(signal.SIGINT, self.signal_handler)
        rospy.loginfo(f"UGV Circle Vel Demo initialized for UGV {self.ugv_id}")

    # 状态回调函数
    def state_callback(self, msg):
        self.current_state = msg

    # 信号处理函数
    def signal_handler(self, signum, frame):
        rospy.loginfo(f"Interrupt signal ({signum}) received. Preparing to return to origin...")
        self.shutdown_requested = True

    # 计算两点间距离
    def distance(self, x1, y1, x2, y2):
        dx = x1 - x2
        dy = y1 - y2
        return math.sqrt(dx * dx + dy * dy)

    # 返回原点函数
    def return_to_origin(self):
        self.returning_to_origin = True
        rospy.loginfo("Returning to origin (0,0)...")

        rate = rospy.Rate(20)
        start_time = rospy.Time.now()

        while not rospy.is_shutdown() and self.shutdown_requested:
            # 检查是否有状态更新
            if self.current_state is None:
                rate.sleep()
                continue

            # 计算当前位置到圆心的距离
            dx = self.current_state.position[0] - self.circle_center_x
            dy = self.current_state.position[1] - self.circle_center_y
            dist = self.distance(self.current_state.position[0], self.current_state.position[1], self.circle_center_x, self.circle_center_y)
            
            #检查是否到达原点
            if dist < 0.1:
                rospy.loginfo("Reached origin (0,0). Stopping...")
                break

            # 检查超时（10秒）
            if (rospy.Time.now() - start_time).to_sec() > 10.0:
                rospy.logwarn("Timeout while returning to origin. Stopping...")
                break

            # 创建并发布速度控制指令
            ugv_cmd = UGVControlCMD()
            ugv_cmd.cmd = UGVControlCMD.POS_CONTROL_ENU
            ugv_cmd.desired_pos[0] = self.circle_center_x
            ugv_cmd.desired_pos[1] = self.circle_center_y
            ugv_cmd.desired_yaw = 0.0

            # 计算速度因子（根据距离调整速度）
            speed_factor = min(1.0, dist / 2.0)
            ugv_cmd.desired_vel[0] = self.linear_speed * speed_factor * (-dx / dist) if dist > 1e-3 else 0.0
            ugv_cmd.desired_vel[1] = self.linear_speed * speed_factor * (-dy / dist) if dist > 1e-3 else 0.0

            self.cmd_pub.publish(ugv_cmd)

            rospy.loginfo_throttle(0.5, f"Distance to origin: {dist:.2f}m")

            rate.sleep()

        # 发布停止指令
        stop_cmd = UGVControlCMD()
        stop_cmd.cmd = UGVControlCMD.HOLD
        self.cmd_pub.publish(stop_cmd)
        rospy.loginfo("UGV stopped at origin.")


    # 生成圆轴运动控制指令
    def generate_commands(self):
        # 检查是否返回原点
        if self.returning_to_origin:
            return
        
        # 创建并发布速度控制指令
        ugv_cmd = UGVControlCMD()
        ugv_cmd.cmd = UGVControlCMD.VEL_CONTROL_ENU

        # 计算当前位置到圆心的距离
        dx = self.current_state.position[0] - self.circle_center_x if self.current_state else 0.0
        dy = self.current_state.position[1] - self.circle_center_y if self.current_state else 0.0
        dist = self.distance(self.current_state.position[0], self.current_state.position[1], self.circle_center_x, self.circle_center_y) if self.current_state else 0.0

        # 处理原点位置或未完成初始移动的情况
        if (dist < 0.05) or not self.initial_move_completed:
            # 初始移动到圆周上
            target_x = self.circle_center_x + self.INITIAL_MOVE_DISTANCE
            target_y = self.circle_center_y

            # 计算到目标点的方向向量
            tx = target_x - self.current_state.position[0] if self.current_state else 0.0
            ty = target_y - self.current_state.position[1] if self.current_state else 0.0
            target_dist = self.distance(self.current_state.position[0], self.current_state.position[1], target_x, target_y) if self.current_state else 0.0
            
            # 归一化向量
            if target_dist > 1e-3:
                tx /= target_dist
                ty /= target_dist

            # 设置速度
            ugv_cmd.desired_vel[0] = self.linear_speed * tx
            ugv_cmd.desired_vel[1] = self.linear_speed * ty

            # 检查是否到达目标点
            if target_dist < 0.1:
                self.initial_move_completed = True
                rospy.loginfo("Initial move completed. Starting circular motion.")
        
    #==================================== 轨迹控制关键代码段 BEGIN（二次开发） ====================================
        # 圆周运动控制
        else:
            # 计算半径误差
            radius_error = dist - self.circle_radius

            # 计算径向单位向量
            radial_x = dx / dist if dist > 1e-3 else 0.0
            radial_y = dy / dist if dist > 1e-3 else 0.0

            # 计算切向单位向量（逆时针方向）
            tangent_x = -radial_y
            tangent_y = radial_x

            # 计算径向校正速度分量（指向圆心）
            v_radial = -self.radial_gain * radius_error

            # 组合速度向量 = 切向速度 + 径向校正速度（指向圆心）
            vx = self.linear_speed * tangent_x + v_radial * radial_x
            vy = self.linear_speed * tangent_y + v_radial * radial_y

            # 限制最大速度
            speed = math.sqrt(vx * vx + vy * vy)
            if speed > 1.5 * self.linear_speed:
                vx *= (1.5 * self.linear_speed) / speed
                vy *= (1.5 * self.linear_speed) / speed
            
            ugv_cmd.desired_vel[0] = vx
            ugv_cmd.desired_vel[1] = vy
    #==================================== 轨迹控制关键代码段 END（二次开发） ======================================

        # 保持当前朝向
        ugv_cmd.desired_yaw = 0.0
        self.cmd_pub.publish(ugv_cmd)

        # 调试信息
        if self.current_state:
            rospy.loginfo_throttle(1.0, f"Position: ({self.current_state.position[0]:.2f}, {self.current_state.position[1]:.2f}), Distance: {dist:.2f}m, State: {'Circular' if self.initial_move_completed else 'Initial Move'}")

    def run(self):
        # 20 Hz循环频率
        rate = rospy.Rate(20)

        # 等待首次状态更新
        rospy.loginfo("Waiting for initial state...")
        while not rospy.is_shutdown() and self.current_state is None:
            rospy.sleep(0.1)

        # 打印初始位置
        rospy.loginfo(f"Starting UGV circular motion controller. Initial position: ({self.current_state.position[0]:.2f}, {self.current_state.position[1]:.2f})")

        # 检查是否需要在原点开始移动
        dist = self.distance(self.current_state.position[0], self.current_state.position[1], self.circle_center_x, self.circle_center_y)

        # 检查是否在圆心位置，若不在则直接开始圆周运动
        if dist < 0.05:
            rospy.loginfo(f"UGV at center. Performing initial move to ({self.INITIAL_MOVE_DISTANCE:.1f}, 0)")
            self.initial_move_completed = False
        else:
            self.initial_move_completed = True
            rospy.loginfo("Starting circular motion directly")
        
        # 主循环，持续生成指令
        while not rospy.is_shutdown() and not self.shutdown_requested:
            self.generate_commands()
            rate.sleep()

        # 程序退出前返回原点
        if not self.returning_to_origin:
            self.return_to_origin()

        # 发布停止指令，确保小车停止
        stop_cmd = UGVControlCMD()
        stop_cmd.cmd = UGVControlCMD.HOLD
        self.cmd_pub.publish(stop_cmd)
        rospy.loginfo("Program exiting.")

def main():
    rospy.init_node('ugv_circular_motion')
    demo = UGVCircleVel()
    demo.run()

if __name__ == '__main__':
    main()

File: References/Sunray/General_Module/sunray_tutorial/ugv_basic/ugv_circle_vel.cpp
#include <ros/ros.h>
#include <sunray_msgs/UGVControlCMD.h>
#include <sunray_msgs/UGVState.h>
#include <cmath>
#include <csignal> // 用于信号处理

// 全局变量声明
ros::Publisher cmd_pub;
sunray_msgs::UGVState current_state;
int ugv_id;
double linear_speed = 0.5;
double circle_radius = 1.0;   // 固定半径为1米
double circle_center_x = 0.0; // 圆心固定为(0,0)
double circle_center_y = 0.0;
double radial_gain = 0.5;                 // 径向校正增益
bool initial_move_completed = false;      // 初始移动完成标志
const double INITIAL_MOVE_DISTANCE = 0.5; // 初始移动距离
bool returning_to_origin = false;         // 返回原点标志
bool shutdown_requested = false;          // 关闭请求标志

// 状态回调函数
void stateCallback(const sunray_msgs::UGVState::ConstPtr &msg)
{
    current_state = *msg;
}

// 计算两点间距离
double distance(double x1, double y1, double x2, double y2)
{
    double dx = x1 - x2;
    double dy = y1 - y2;
    return std::sqrt(dx * dx + dy * dy);
}

// 信号处理函数 (用于捕获Ctrl+C)
void signalHandler(int signum)
{
    ROS_INFO("Interrupt signal (%d) received. Preparing to return to origin...", signum);
    shutdown_requested = true;
}

// 返回原点函数
void return_to_origin()
{
    returning_to_origin = true;
    ROS_INFO("Returning to origin (0,0)...");

    ros::Rate rate(20);
    ros::Time start_time = ros::Time::now();

    while (ros::ok() && shutdown_requested)
    {
        double dx = current_state.position[0] - circle_center_x;
        double dy = current_state.position[1] - circle_center_y;
        double dist = distance(current_state.position[0], current_state.position[1],
                               circle_center_x, circle_center_y);

        // 检查是否到达原点
        if (dist < 0.1)
        {
            ROS_INFO("Reached origin (0,0). Stopping...");
            break;
        }

        // 检查超时（10秒）
        if ((ros::Time::now() - start_time).toSec() > 10.0)
        {
            ROS_WARN("Timeout while returning to origin. Stopping...");
            break;
        }

        // 创建位置控制命令
        sunray_msgs::UGVControlCMD ugv_cmd;
        ugv_cmd.cmd = sunray_msgs::UGVControlCMD::POS_CONTROL_ENU;
        ugv_cmd.desired_pos[0] = circle_center_x;
        ugv_cmd.desired_pos[1] = circle_center_y;
        ugv_cmd.desired_yaw = 0.0;

        // 设置速度（根据距离调整）
        double speed_factor = std::min(1.0, dist / 2.0);
        ugv_cmd.desired_vel[0] = linear_speed * speed_factor * (-dx / dist);
        ugv_cmd.desired_vel[1] = linear_speed * speed_factor * (-dy / dist);

        cmd_pub.publish(ugv_cmd);

        ROS_INFO_THROTTLE(0.5, "Distance to origin: %.2fm", dist);

        ros::spinOnce();
        rate.sleep();
    }

    // 发送停止指令
    sunray_msgs::UGVControlCMD stop_cmd;
    stop_cmd.cmd = sunray_msgs::UGVControlCMD::HOLD;
    cmd_pub.publish(stop_cmd);
    ROS_INFO("UGV stopped at origin.");
}

// 生成圆周运动控制命令
void generate_commands()
{
    sunray_msgs::UGVControlCMD ugv_cmd;

    // 处理返回原点请求
    if (returning_to_origin)
    {
        return; // 不生成命令，由return_to_origin函数处理
    }

    ugv_cmd.cmd = sunray_msgs::UGVControlCMD::VEL_CONTROL_ENU;

    // 计算当前位置与圆心的距离
    double dx = current_state.position[0] - circle_center_x;
    double dy = current_state.position[1] - circle_center_y;
    double dist = distance(current_state.position[0], current_state.position[1],
                           circle_center_x, circle_center_y);

    // 处理原点位置或未完成初始移动的情况
    if ((dist < 0.05) || !initial_move_completed)
    {
        // 初始移动阶段：移动到(0.5, 0)位置
        double target_x = circle_center_x + INITIAL_MOVE_DISTANCE;
        double target_y = circle_center_y;

        // 计算到目标点的向量
        double tx = target_x - current_state.position[0];
        double ty = target_y - current_state.position[1];
        double target_dist = distance(current_state.position[0], current_state.position[1],
                                      target_x, target_y);

        // 归一化向量
        if (target_dist > 1e-3)
        {
            tx /= target_dist;
            ty /= target_dist;
        }

        // 设置速度
        ugv_cmd.desired_vel[0] = linear_speed * tx;
        ugv_cmd.desired_vel[1] = linear_speed * ty;

        // 检查是否到达初始位置
        if (target_dist < 0.1)
        {
            initial_move_completed = true;
            ROS_INFO("Initial move completed. Starting circular motion.");
        }
    }

    //==================================== 轨迹控制关键代码段 BEGIN（二次开发） ====================================
    else
    {
        // 圆周运动阶段
        // 计算半径误差
        double radius_error = dist - circle_radius;

        // 计算径向单位向量
        double radial_x = 0.0, radial_y = 0.0;
        if (dist > 1e-3)
        {
            radial_x = dx / dist;
            radial_y = dy / dist;
        }

        // 计算切向单位向量 (逆时针方向)
        double tangent_x = -radial_y;
        double tangent_y = radial_x;

        // 计算径向校正速度分量 (指向圆心)
        double v_radial = -radial_gain * radius_error;

        // 组合速度向量 = 切向速度 + 径向校正速度
        double vx = linear_speed * tangent_x + v_radial * radial_x;
        double vy = linear_speed * tangent_y + v_radial * radial_y;

        // 限制最大速度
        double speed = std::sqrt(vx * vx + vy * vy);
        if (speed > 1.5 * linear_speed)
        {
            vx *= (1.5 * linear_speed) / speed;
            vy *= (1.5 * linear_speed) / speed;
        }

        ugv_cmd.desired_vel[0] = vx;
        ugv_cmd.desired_vel[1] = vy;
    }
    //==================================== 轨迹控制关键代码段 END ====================================
    
    ugv_cmd.desired_yaw = 0.0; // 保持当前朝向
    cmd_pub.publish(ugv_cmd);

    // 调试信息
    static ros::Time last_log_time = ros::Time::now();
    if ((ros::Time::now() - last_log_time).toSec() > 1.0)
    {
        ROS_INFO("Position: (%.2f, %.2f), Distance: %.2fm, State: %s",
                 current_state.position[0], current_state.position[1],
                 dist, initial_move_completed ? "Circular" : "Initial Move");
        last_log_time = ros::Time::now();
    }
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "ugv_circular_motion");
    ros::NodeHandle nh("~");

    // 注册信号处理函数
    signal(SIGINT, signalHandler);

    // 获取参数
    nh.param<int>("ugv_id", ugv_id, 1);
    nh.param<double>("linear_speed", linear_speed, 0.5);
    nh.param<double>("radial_gain", radial_gain, 0.5);

    // 创建发布器和订阅器
    std::string cmd_topic = "/ugv" + std::to_string(ugv_id) + "/sunray_ugv/ugv_control_cmd";
    std::string state_topic = "/ugv" + std::to_string(ugv_id) + "/sunray_ugv/ugv_state";

    cmd_pub = nh.advertise<sunray_msgs::UGVControlCMD>(cmd_topic, 10);
    ros::Subscriber state_sub = nh.subscribe(state_topic, 1, stateCallback);

    // 等待第一次位置更新
    ROS_INFO("Waiting for initial state...");
    ros::Duration(1.0).sleep();
    ros::spinOnce();

    if (ros::ok())
    {
        ROS_INFO("Starting UGV circular motion controller");
        ROS_INFO("Initial position: (%.2f, %.2f)",
                 current_state.position[0], current_state.position[1]);

        // 检查是否需要在原点开始移动
        double dist = distance(current_state.position[0], current_state.position[1],
                               circle_center_x, circle_center_y);
        if (dist < 0.05)
        {
            ROS_INFO("UGV at center. Performing initial move to (%.1f, 0)", INITIAL_MOVE_DISTANCE);
            initial_move_completed = false;
        }
        else
        {
            initial_move_completed = true;
            ROS_INFO("Starting circular motion directly");
        }

        ros::Rate rate(20); // 20Hz控制频率
        while (ros::ok() && !shutdown_requested)
        {
            generate_commands();
            ros::spinOnce();
            rate.sleep();
        }

        // 程序退出前返回原点
        if (!returning_to_origin)
        {
            return_to_origin();
        }
    }

    // 确保小车停止
    sunray_msgs::UGVControlCMD stop_cmd;
    stop_cmd.cmd = sunray_msgs::UGVControlCMD::HOLD;
    cmd_pub.publish(stop_cmd);
    ROS_INFO("Program exiting.");

    return 0;
}